### ISSUE TITLE
`ssrc` in `RTCRtpEncodingParameters` is inconsistent with ORTC

### ISSUE THREAD

**Issue by taylor-b** [w3c/webrtc-pc#1174]
My assumption is that `sender.getParameters` will return the SSRC identifier that's currently being used for sending. It's described as:

> The SSRC of the RTP source stream of this encoding (non-RTX, non-FEC RTP stream).

However, in ORTC it's a bit more complex, and an unset `ssrc` has a special meaning:

> If ssrc is unset in a RTCRtpEncodingParameters object passed to the send() method, the browser will choose, and the chosen value is not reflected in RTCRtpEncodingParameters.ssrc. If the browser chooses the ssrc, it may change it due to a collision without firing an RTCSsrcConflictEvent. If ssrc is set in an RTCRtpEncodingParameters object passed to the send() method and an SSRC conflict is detected within the RTP session, then an RTCSsrcConflictEvent is fired (see Section 5.4).

So basically: when `ssrc` is set, the application must explicitly handle SSRC conflicts by listening to an event. But when it's unset, it's like a wildcard, and can change automatically due to SSRC conflicts.

WebRTC 1.0 always behaves in the latter way. So, if the ORTC API were mapped directly to WebRTC, the `ssrc` field would always be unset.

Does WebRTC really need to expose the current SSRC here? What is it expected to be used for?

Also, one very corner-casey problem that could occur if `getParameters` returns the current SSRC:

1. `getParameters` called, returns SSRC A.
2. SSRC conflict suddenly occurs, it's changed to SSRC B.
3. `setParameters` called with modified parameters. Fails with `InvalidModificationError` since it's interpreted as trying to change the SSRC from B to A.

**Comment by aboba** [comment]
How do we think the SSRCs returned by getParameters() will be used?  This includes not only encodings[].ssrc, but also encodings[].rtx.src and encodings[].rtx.fec.ssrc.  Since they can't be changed using setParameters(),  should we consider this more appropriate for webrtc-statistics? 

The other alternative would be to ignore attempts to set SSRCs in setParameters() and addTransceiver(), rather than failing with `InvalidModificationError`.

@fippo What do you think?

**Comment by fippo** [comment]
I don't think they will be very useful... ignoring attempts to set them SGTM while not creating too much hassle for ORTC compat

I wondered about the stats case too. https://w3c.github.io/webrtc-stats/#streamstats-dict* has a ssrc where the ssrc may change in a subsequent call to getStats. That _might_ allow detecting a collision.

**Comment by aboba** [comment]
@fippo @taylor-b Unfortunately, we cannot only guarantee that `sender.getParameters()` returns the SSRC used for sending at the instant that the method returns. By the time `sender.setParameters` is called, the SSRC could have changed.  The point of adding `transactionId` to `RTCRtpEncodingParameters` was to make sure that there were no intervening changes between `sender.getParameters()`  and `sender.setParameters()` but if the SSRCs can change, this is impossible.

**Comment by aboba** [comment]
From https://www.w3.org/2017/06/28-webrtc-minutes 

Taylor to do a PR to remove the odd behavior

**Comment by aboba** [comment]
I think this needs to be reopened because most existing SFUs use SSRC signaling.  So if the ssrcs aren't available from SDP, they need to be made available in the encoding parameters.  Removing the ssrc information will in practice block the transition away from "Plan B".

**Comment by aboba** [comment]
Consensus from the April VI is to develop a PR re-introducing SSRCs into `RTCRtpEncodingParameters`.

**Comment by taylor-b** [comment]
@aboba Where does this currently stand? When we had a call talking about this, I recall deciding that it would be simpler for browsers to just continue putting SSRCs in SDP while the ecosystem transitions to MID/RID-based demuxing.

**Comment by aboba** [comment]
Since it appears that Unified Plan implementations are intending to include "a=ssrc" lines, the information should be available for existing SFUs to function without any changes to the WebRTC 1.0 specification.  I am therefore closing this issue.  If you believe this is an error, please reopen.

**Comment by aboba** [comment]
Re-opening, because Unified Plan implementations are not all including "a=ssrc" lines as was agreed upon to close this issue.

**Comment by emcho** [comment]
I was really quite relieved when we discussed this with Taylor and Bernard and agreed that SSRCs will be in the SDP. Looking at the recent Chrome 74 implementation is making me rather nervous though as there are no SSRCs in sight and this effectively prevents us from upgrading the Jitsi ecosystem to use the new standard simulcast support. Jitsi Videobridge routes exclusively based on SSRCs
Could we please have this explicitly stated in the spec?

**Comment by alvestrand** [comment]
@emcho can you post a link to a demo? I'd like to check the actual SDP produced.

**Comment by lminiero** [comment]
I'm on `74.0.3717.0` and ssrcs seem to still be there, not sure if they started disappearing in more recent versions instead?

**Comment by murillo128** [comment]
The main problem of including SSRCS on the SDP is that there is no standard way to assign each SSRC to the corresponding RID value. BTW, when mangling the SDP (pre-74) to enable simulcast I signal it by adding a custom parameter to the rid attribute.

```
a=rid:a send ssrc=1234;
```

**Comment by murillo128** [comment]
@lminiero i think you need to enable simucast via `addTransceiver`

Here you have a fiddle for it:
https://jsfiddle.net/nd8ujcpr/7/

By the way, `rids` are still ignored in chrome 74, neither on the SDP nor in the `getParameters`

**Comment by lminiero** [comment]
Oh, didn't realize this was a simulcast-specific issue, I thought SSRCs had gone altogether. Thanks for the clarification!

**Comment by murillo128** [comment]
@emcho is there any technical reason preventing jitsi to support signaling with `rid` instead of with `ssrc` or is it a matter of the development cost of the update?

**Comment by nils-ohlmeier** [comment]
How about the SFU's implementing support for MID and RID and learn the SSRC's from the RTP packets like the spec says to do?
I think if we keep SSRC around for "legacy" implementations then nobody is ever going to switch to the SSRC-free world which MID and RID promised, because everyone is going to be lazy and use the SSRCs as they are still available.

**Comment by aboba** [comment]
@nils-ohlmeier The "ssrc-free world" depends on MID header extension implementation, because MIDs are supported in the routing rules (RIDs are not).  Unfortunately, full MID support (including in SDES RTCP packets) has lagged behind.  So we're in a situation where we want applications to migrate but aren't providing all the tools to allow that to happen quickly.

**Comment by amithilbuch** [comment]
This is my personal opinion and is not the official opinion of my employer.

I think i need to understand the "backwards-compatibility" motivation a bit more.
The way i see it, this is new functionality and a "new spec" in unified plan. Work needs to be done in SFUs in order to support it, as we can't expect the SFUs to understand the a=simulcast attribute and understand what it needs to signal it back to acknowledge receipt of simulcast. So this won't work out of the gate even if we were to include SSRCs.
On the other hand, I feel that it is reasonable to assume that an SFU that understands a=simulcast should also understand RIDs and MIDs, as those specs are in the dependency chain for simulcast.

It is worth noting that  we did not take away the legacy simulcast munging scenario, and that can still be used. Nor did we remove SSRCs from any existing supported scenario, only in the spec-compliant simulcast scenario where SSRCs "conflict" with RIDs as the source of truth regarding layer identification.

I also feel that continuing to push to a hybrid of plan-b and unified-plan specs (which is what is happening de facto, with all the support that gets baked into implementations) hurts us in the long run, as it gives more excuses for applications not to migrate (they keep seeing new features), while also creating unnecessarily complicated (and therefore buggy) software. In the end, it leads to the opposite of what we want, as i have seem multiple spec scenarios that won't work as intended because implementations also require parts borrowed from other specs, as is evident by some of the things that break when SSRCs are not signaled.

**Comment by amithilbuch** [comment]
@murillo128 Regarding RIDs being ignored in chrome. 
This is currently by design, as there is no need for them:
1. It has been made clear in the past (in the last interim as well), that the simulcast scenario requires an SFU, and that PC does not support receiving simulcast. Two peers do not need to use simulcast because they can directly negotiate the best configuration for their session without sending multiple layers. In group call scenario, not having an SFU means that we have a quadratic number of connections and that leads to a terrible experience that would be worsened by trying to send and receive multiple layers.
As such no support has been added to PC to enable receiving simulcast. The [spec](https://rawgit.com/w3c/webrtc-pc/master/webrtc.html#simulcast-functionality) also indicates that there is only support for issuing a=simulcast with the send direction, consistent with this decision. To be fair it also mentions the scenario of receiving but does not define how this should be done.
2. Without simulcast, the use of RID is redundant as MIDs are sufficient identifiers.
Assuming unified-plan semantics, MID is sufficient to identify streams and route them correctly so RIDs are redundant in this sense. Moreover, the [spec](https://rawgit.com/w3c/webrtc-pc/master/webrtc.html#dom-rtcpeerconnection-addtransceiver) indicates that a single RID should be removed and is not negotiated in the SDP:
`
If the number of RTCRtpEncodingParameters now stored in sendEncodings is 1, then remove any rid member from the lone entry.
`

**Comment by aboba** [comment]
@amithilbuch  The original resolution of this issue was proposed by Taylor Brandstetter:

"I recall deciding that it would be simpler for browsers to just continue putting SSRCs in SDP while the ecosystem transitions to MID/RID-based demuxing."

Based on acceptance of Taylor's proposal, the WEBRTC WG removed SSRCs from RTCRtpEncodingParameters. 

If there is no longer consensus on this, then we will need to revert the SSRC removal PR.  Are you saying that you would prefer that approach?

**Comment by murillo128** [comment]
@amithilbuch my use case is adding a transceiver with multiple encodings for enabling simulcast to talk to an SFU:

```js
pc.addTransceiver(videoTrack,{
	direction	: "sendonly",
	streams		: [stream],
	sendEncodings	: [
        	{"rid":"a"},
                {"rid":"b","scaleDownResolutionBy":2},	
                {"rid":"c","scaleDownResolutionBy":4}
        ]
});
```
`rid` is not set correcty and `pc.getTransceivers()[0].sender.getParameters().encodings` values are not set:
```
[{"active":true,"networkPriority":"low","priority":"low"},
{"active":true,"networkPriority":"low","priority":"low"},
{"active":true,"networkPriority":"low","priority":"low"}]
```
and in SDP are also not set:
```
a=rid:0 send
a=rid:1 send
a=rid:2 send
a=simulcast:send 0;1;2
```
There is a CL already that supposedly implements this and I hope it lands before getting to beta.

**Comment by amithilbuch** [comment]
@aboba I think we should look at this in the context of the two separate scenarios.
SSRCs are signaled in the SDP in all scenarios except in the spec-compliant simulcast. Unified plan can still be used for simulcast with SSRCs with SDP munging. As such i do not think that we should revert that PR.

In the spec-compliant simulcast scenario, we require the endpoint to respond with `a=simulcast`, `a=rid` and acknowledge the rid header extension, but we are not requiring it to actually stand behind what signaling these attributes means? If the endpoint does not signal `a=rid` or `a=simulcast` simulcast cannot be used and the PC will revert to sending only the first layer.

The alternative would be to introduce something like @murillo128 suggested `a=rid:a send ssrc=123`. But this would be new SDP grammar. It seems that unless the SDP is munged to add `a=rid` and `a=simulcast`, changes have to be made to SFUs to support the spec-compliant scenario.  If it is the intent of owners of these endpoints to munge the SDP to achieve this, then there is a supported legacy scenario that already covers the behavior.

**Comment by amithilbuch** [comment]
@murillo128 RIDs are not wired into WebRTC yet, this is waiting for API owners LGTM so that it can be merged. It is expected to land before beta.
RIDs are currently auto-generated by the platform because they were not passed down by the browser.
You are not seeing the scale down by values because you are not matching the [idl](https://cs.chromium.org/chromium/src/third_party/blink/renderer/modules/peerconnection/rtc_rtp_encoding_parameters.idl?l=23).
try `"scaleResolutionDownBy": 2`

**Comment by murillo128** [comment]
> Unified plan can still be used for simulcast with SSRCs with SDP munging. As such i do not think that we should revert that PR.

Do you know how much time are you going to support both the SDP munging and the standard way of enabling simulcast? That should provide enough time for people to upgrade their SFUs to use rids.

Re: `scaleResolutionDownBy` yes it is a typo on the jsfiddle, thank you for pointing it out.

**Comment by amithilbuch** [comment]
I am not aware of a timeline to stop supporting the munging scenario. 
If a date is determined i am fully confident that it will be communicated broadly, will give users enough time to prepare and make a transition, and that the community will be consulted for preferences on any key dates.
I think the recent transition to Unified Plan by default is a good example of how such a change happens.

**Comment by fippo** [comment]
FWIW: I think the unified plan transition was a desaster in terms of (lack of QA)

**Comment by emcho** [comment]
@alvestrand i was about to post another fiddle but the one from Sergio above demonstrates the issue very well.

**Comment by amithilbuch** [comment]
@fippo your feedback is worth a lot, but i do not think this thread is a good place for a discussion on the unified plan transition.
Feel free to shoot me an email or to open a thread and tag me (but i honestly don't know where) to discuss any input you might have on that transition so that we can improve in the next one.

**Comment by fippo** [comment]
@amithilbuch: @henbos knows my opinion on this at length and can give you the gist.

This is somewhat relevant to this discussion since in non-simulcast Chrome has opted to include a=ssrc lines for practical reasons instead of going for what unified plan attempts to specify. I saw this as an indicator that https://github.com/w3c/webrtc-pc/issues/1174#issuecomment-398525608 is the consensus. Like the codec consensus nobody was happy with that solution but...

**Comment by amithilbuch** [comment]
@aboba I think the discussion and original consensus is on a much broader subject - unified plan scenario as a whole and chrome has lived up to that and signaled SSRCs for all of the existing backwards compatible scenarios.

If our goal is to have SSRCs everywhere (both new and existing apis), then we should consider putting it explicitly in the spec. And we might as well standardize `a=ssrc-group:SIM` in the SDP instead of the `a=simulcast` attribute to achieve the same. 
I'm glad you tagged this issue with March interim as we should have a broad discussion.

This is a new API surface and it might not fall under the same rules because the semantics are different (no existing clients, mandates SFUs, RIDs must be negotiated, etc.). We might want to keep this scenario exempt from signaling these SSRCs because without directly tying them to RIDs it could be destructive. Here are a few examples of things that could go wrong:

imagine i want to send you 3 simulcast layers and you want to accept only two (you reject one layer).
how would you know which ssrc is the one you should remove along with the RID? what if you reject simulcast altogether, how do you know which of the 3 ssrcs that were proposed is the one that will be used in the single layer that eventually gets sent?

Simulcast needs to be acknowledged by the receiving party, that means that the ssrcs from the sender will need to be in the answer along with the ssrc for video sent from the SFU on the same section. This will be different than any existing scenario (SSRCs are never acked).

RTX, flex, or any other SSRCs that might need to be signaled along on the same section would cause more confusion on which SSRCs are associated with which streams. how would the implementation know how to disambiguate them as well?

The simulcast receiver (SFU) could be the one initiating the call using `a=simulcast:recv 1;2;3`.
Should it also indicate the ssrcs that correspond to each rid? that would be different than any existing scenario.

**Comment by aboba** [comment]
@murillo128 SSRCs were formerly provided in `RTCRtpEncodingParameters`, as read-only parameters, along with `rid`.  The question arose as to what would happen in the situation where `ssrc` changed in between `get` and `setParameters`.  This didn't seem intractable (one option is to ignore the attempt to set `ssrc`, rather than error), but Taylor's take was it would be easier to include SSRCs in the Unified Plan SDP for simulcast, similar to how it was done for non-simulcast scenarios. However, this did not take into account the need to relate RIDs and SSRCs in SDP, something which the object model addressed.  Can you describe the scenario for needing to relate them?

@amithilbuch  If SSRCs are not provided in the object model or in the SDP, how is a simulcast application to obtain SSRCs in order to signal to a legacy SFU?  Not sure what you mean by "SDP munging" in that scenario. 

@fippo One potential impact of removing SSRC info is the effect on simulcast loopback testing in WPT (e.g. "simulcast playground").  It's not possible to build "simulcast playground" with MIDs or RIDs alone, right?

**Comment by fippo** [comment]
@aboba the demultiplexing approach on the receiver side is a theoretically merely more complicated as, lacking ssrc in RTCRtpEncodingParameters on the sender side (to stay on-topic) the only way to retreive them is getStats on the receiver side (sender side stats for simulcast are busted in Chrome) followed by renegotiation to pull those tracks out of the peerconnection.

In practice I seem to manage to confuse the decoder enough to lock up my machine and spec-stats do not seem to return those ssrcs :joy:

**Comment by murillo128** [comment]
@aboba IMHO the final goal should be to be compliant with the specs (i.e. `rid` only) while giving a transition path for all the SFUs to migrate to it without nasty surprises. In that regards chrome M74 does not break compatibility with any implementation as the old SDP mungling way of enabling simulcast is still supported.

**Comment by murillo128** [comment]
@aboba @fippo according to bundle rfc, the simulcast playground should not work as `mid` has preference over `ssrc`

```
      If the MID associated with the RTP stream is not in the table
      mapping MID to "m=" section, then the RTP stream is not decoded
      and the payload data is discarded.
```

**Comment by murillo128** [comment]
@aboba @fippo I (partially) correct myself, I assume that you are not negotiating the `mid` extension on the SDP, so it would be ignored by the receiver and only do the mapping on the `ssrc` value. Not sure that you can prevent `mid` association via RTCP SDES packets.

**Comment by fippo** [comment]
@murillo128 correct, making the sender think its ok to send mid while making the receiver ignore the extension is the way to go about this.

**Comment by lminiero** [comment]
@fippo that's true in a generic SFU that has control over all the aspects of the media routing, which unfortunately is not my case. The SFU part in my case is a module, so the media routing and the WebRTC termination part are separated, and the server won't know where media will go to or come from. As such, there could be sendrecv connections, where accepting the extensions means accepting them in all directions: since the source of the media may be a different PeerConnection with completely different MID mappings, I might end up sending the user an RTP packet with mid=x where it's expecting mid=y. For this reason, at the moment I'm rejecting it on the way in too, and just rely on SSRC.

I know this isn't really a spec issue, or an issue any of you should be really be bothered about, but it was just to clarify that saying "that's what SFUs should do" may not be practical in some cases. I guess eventually I'll have to bite the bullet and just always accept the mid extension, and generate my own extension on the way out (which I'd rather not do for performance reason but whatever).

**Comment by murillo128** [comment]
@lminiero, i think @fippo was only talking about the simulcast playground on WPT.
 
Re. your architecture, I have it also like that, but I do an initial `mid` to `ssrc` mapping at the receiver side and work internally with `ssrc` based on that. You will only need to insert the `mid` (if negotiated) accordingly as you do the normal `ssrc` rewriting  when sending to the remote party.

**Comment by amithilbuch** [comment]
@aboba I am against SDP munging in applications overall. 
What i meant is that if the (legacy) SFU doesn't support RIDs, then the application will have to munge the SDP to simulate the `a=rid` and `a=simulcast` lines arriving from the SFU while negotiating a completely different SDP to the SFU - one that probably closely resembles `a=ssrc-group:SIM`.
And in that case, the application is better off* using the legacy munging technique to achieve simulcast, without the new API. We are still supporting it.
*Better off = will involve less munging (only one way) and less bugs (no need to migrate code).

I suggest that users move to the new API when they are ready to accept a munge-free lifestyle.

**Comment by emcho** [comment]
@amithilbuch I am still not clear how you see it will be possible, (munging or no munging), to provide the SFU with SSRCs ... since they are completely gone now.

**Comment by amithilbuch** [comment]
@emcho it is currently not possible to extract the SSRCs.
I was making a hypothetical argument, that if you intend on munging the SDP, then you should continue using the legacy technique. It is still supported.
We do not want people to munge the SDP with the new API. You should be using it in the spec-compliant way.

**Comment by aboba** [comment]
@emcho As referred to in this issue, the WebRTC 1.0 APIs formerly included SSRCs in the object model so that applications could signal them in the Offer, regardless of the content of the SDP.  The functionality was removed based on Taylor's assertion that putting the SSRC information into the SDP would reduce implementation time for browsers and SFUs.

**Comment by emcho** [comment]
@amithilbuch indeed, and this discussion is about defining what the spec-compliant way is. Specs get written with various criteria in mind. One important piece to always take into consideration is the current state of the art. There has been an omission on that so far and this is what we are trying to fill in.

@aboba yes, certainly! Any place would work ... as long as it is actually somewhere.

**Comment by fippo** [comment]
So I made the "playground approach" work, see [here](https://github.com/fippo/simulcast-playground/blob/gh-pages/spec.html)
It doesn't extract the tracks in resolution order but that was a weak assertion anyway. And we get the decoded images so can query their resolution. There are some rough edges such as relying on plan-b in Chrome in order to generate an answer that is mid- and rid-less but this  should still work with unified plan with slight modifications (which are already in the firefox code)

Ironically Chrome misinterpreted a=msid: without any a=ssrc:msid lines and got confused by the addition of a=ssrc:... msid later. I think Taylor wrote that :-)

**Comment by ibc** [comment]
Hi, I was about to write here about a related issue, but finally did it in a separate issue:

"Missing CNAME in RID based simulcast offer": #2114

**Comment by ibc** [comment]
> Re. your architecture, I have it also like that, but I do an initial `mid` to `ssrc` mapping at the receiver side and work internally with `ssrc` based on that. You will only need to insert the `mid` (if negotiated) accordingly as you do the normal `ssrc` rewriting when sending to the remote party.

This is not so easy. In my experiments (don't remember now which browser/version):

* If `a=extmap:3 urn:ietf:params:rtp-hdrext:sdes:mid` is negotiated in the PeerConnection and a RTP packet with a header extension with `id=3` is received, and assuming that the remote SDP signals `a=ssrc` lines:
   * The browser will drop the packet if its SSRC matches the one in `a=ssrc` but its RTP MID does not match the `a=mid` of the corresponding m= section.

* If the packet does not contain RTP MID extension, I don't remember what happens (I hope the browser does not drop it since, according to the spec, RTP header extensions do not need to be included in all packets).

In a SFU scenario sometimes you route packets produced by an endpoint that does not include RTP MID extension into them, or MID was not just negotiated in its signaling leg. If present, the SFU may "just" rewrite the MID `id` in the packet so the receiver does not try to match via MID. Rewriting the MID value itself is tedious since it has dynamic size and mangling a RTP packet to include/extend/delete a header extension is not something "cool" at all.

In summary, this is tricky. I'm super ok with browsers not signaling their SSRCs in the offer because having MID and RID properly implemented in the SFU is the way to go. However I agree that, in the other direction when the SFU sends packets to receivers, this is much more error prune. I avoid having PeerConnections to both send and receive, so I just negotiate MID in sending PeerConnections. But people using bidirectional PeerConnection may encounter many problems as described above.

**Comment by lminiero** [comment]
@ibc just summarized my main concerns here. In my case, my server is not a pure SFU, so while there may be scenarios with sendonly/recvonly PeerConnections, there will also be sendrecv PeerConnections hooked together where the respective MIDs may not match. Which means that regenerating RTP extensions on the way out would be the only way to include the proper MID there, something that as Iñaki pointed out is quite an annoying (but I guess at this point unavoidable) procedure.

**Comment by murillo128** [comment]
> Rewriting the MID value itself is tedious since it has dynamic size and mangling a RTP packet to include/extend/delete a header extension is not something "cool" at all.

I have to disagree here, you will have to add headers (transportwide-cc,abs-send-time) and remove header extensions (client-to-mixer-audio-level/rids) negotiated between SFU and peers anyway, or change the ids (from 1 byte to 2 byte headers). If you expect to do a zero copy of the in buffer to the output buffer I think you are doing it wrong IMHO.

**Comment by ibc** [comment]
> I have to disagree here, you will have to add headers (transportwide-cc,abs-send-time) and remove header extensions (client-to-mixer-audio-level/rids) negotiated between SFU and peers anyway, or change the ids (from 1 byte to 2 byte headers). If you expect to do a zero copy of the in buffer to the output buffer I think you are doing it wrong IMHO.

Well, of course I'll do. But the point here is that people is complaining because the "new" MID/RID mechanism breaks their working SFU logic (that depend on signaled `a=ssrc` values) even if Firefox implements it for long. Forcing them to also implement RTP header extension rewriting/addition/removal do not seem a comfortable solution. First of all, because they may not need it if they don't implement `transport-cc` and happily assume 1-byte heder extensions.

Also, removing client-to-mixer-audio-level/rids header extensions is not needed. It's easier to just not negotiate them at SDP level with the receivers.

**Comment by murillo128** [comment]
What we can't do is specify one thing and implement another. Providing access to the `ssrc` values in `getParameters` and a clean migration path is something we should aim, but claiming that rewriting a RTP header extension on server side is something hard or annoying.. well..

**Comment by ibc** [comment]
Rewriting a RTP header extension (both its `id` and `value`) on server side is not hard. Addition/removal/value-expansion of RTP header extensions is harder. Not sure which your position in this subject is other than making it clear that you have implemented that feature in your stuff and that any other approach is wrong.

**Comment by murillo128** [comment]
My position is that changing the spec because rtp header extension rewriting is required on server side and it is considered difficult is not a valid point IMHO and this thread has evolved from an specific issue about how to handle `ssrc` on the API to an MID/RID vs ssrc discussion that I don't think is the correct place for having it.

**Comment by ibc** [comment]
> thread has evolved from an specific issue about how to handle ssrc on the API to an MID/RID vs ssrc discussion that I don't think is the correct place for having it.

The thing here is that, due to SDP O/A negotiation, it's currently "easy" to implement proper MID/RID parsing and correlation in server side, but if the PeerConnection is used for bidirectional media then complex RTP header extensions mangling is required in server side for transmitting media to receivers in those bidirectional pcs.

**Comment by lminiero** [comment]
Apologies for the possibly out of scope message, but following back on a small exchange we had during the interim call (I didn't know how to get in touch with Amit offline), this is just to confirm to @amithilbuch that I do get proper rid RTP extensions in my packets now using M74 simulcast: I was negotiating back using `a=simulcast:recv rid=xxx` instead of `a=simulcast:recv xxx`, so fixing that fixed the missing simulcast support.

As a side note, though, the former is what Firefox currently does (just tested with Nightly), so Firefox developers may want to fix that, as Chrome seems to be doing the right thing according to [the spec](https://tools.ietf.org/html/draft-ietf-mmusic-sdp-simulcast-14).

**Comment by ibc** [comment]
Firefox still uses the simulcast draft version 03.

**Comment by aboba** [comment]
At the March 2019 interim, there was no interest in changing the WebRTC object model to surface SSRCs.  The debate centered around whether "a=ssrc" lines should be included in SDP - but that is an issue for the IETF rather than the W3C.  As a result, this issue is being closed.
